### DISCUSSION TITLE
Exception message to client differs when running from docker

### DISCUSSION THREAD

**Discussion by abelal83** [serenity-is/Serenity#6409] [General]
Hi all, 

I have this very peculiar issue which I can't figure out the cause.

In Repository.cs I have a condition to throw exception

```csharp
// shortened snippet
protected override void BeforeSave() {
 base.BeforeSave();
.....
 if (isSqlServerPrimary == false) {
throw new ApplicationException($" Request for the SQL Server {sqlServers.Entities[0].Name} is not possible via Serene beause the server is a secondary replica in readonly mode. Please contact Platform Engineering or DBAs for access.");
}
.....
}
```

Fairly straight forward bit of code.

Error message to client when running locally
![image](https://user-images.githubusercontent.com/20558784/171562581-39fd672d-a905-4011-9293-c07e60de3c07.png)

Where it gets peculiar is when the exact same files used to run it locally are oved into a docker container... 

Error message to client when running from docker
![image](https://user-images.githubusercontent.com/20558784/171561830-c956b1a8-ec8f-4547-9215-0c6a78ed0cfa.png)

It doesn't seem to make sense, something appears to be swallowing the exception and retuning a generic error message but ONLY when running inside docker.

Anyone come across this before or have a clue what's going on?

Thanks all in advance.

**Comment by edwardch** [comment]
In docker it might be a networking issue, or something else completely... 

Can you confirm that it is breaking on that exact same line?

Maybe you can take a look at your exception log, or run `docker service logs`, or add some code to do something that confirms your code has reached this point.

With docker, logs are your friend!

**Comment by abelal83** [comment]
Thanks @edwardch for your response!

I should have added we did add a lot of logging for exactly what you mentioned. And yes it is definitely hitting that line

```csharp
if (isSqlServerPrimary == false) {
  Console.WriteLine("ValidateServerIsAllowedForRequest: IsSqlServerPrimary is false");

  throw new ApplicationException($" Request for the SQL Server {sqlServers.Entities[0].Name} is not possible via Serene beause the server is a secondary replica in readonly mode. Please contact Platform Engineering or DBAs for access.");
}
```

We see 'ValidateServerIsAllowedForRequest: IsSqlServerPrimary is false' logged both locally and in docker

Below is the output from both, exactly the same.
```log
Execution for ValidateServerIsAllowedForRequest completed
ValidateServerIsAllowedForRequest: IsSqlServerPrimary is false
info: Microsoft.AspNetCore.Mvc.StatusCodeResult[1]
      Executing HttpStatusCodeResult, setting HTTP status code 500
info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[2]
      Executed action Serene.Azure.Endpoints.SqlDbRequestController.Create (Serene.Web) in 22771.856ms
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'Serene.Azure.Endpoints.SqlDbRequestController.Create (Serene.Web)'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 23679.3192ms 500 application/json
```

It's really strange, I'm copying the exact same compiled files used to run locally to the docker image. 

```docker
FROM mcr.microsoft.com/dotnet/aspnet:3.1
WORKDIR /app
COPY /bin/Release/netcoreapp3.1/publish .
EXPOSE 5000/tcp
ENV ASPNETCORE_URLS http://*:5000
ENTRYPOINT ["dotnet", "Serene.Web.dll"]
```

I can't quite put my finger on it other than assume 'something' in docker image is swallowing and then rethrowing generic error.

This issue actually originated from deployed image to our kubernetes cluster which lead me to test locally with docker, the same issue persists. It was then further tested on another desktop and again same behavior, so that's 3 different machines, 1 Ubuntu server running Kubernetes, and 2 desktops running Docker.

**Reply by edwardch** [reply]
Damn. Think we just need to debug _harder_ :) 

A couple of things I can think of trying:

- You use `{sqlServers.Entities[0].Name}`, can you confirm all set correctly by doing a writeline in a foreach on the `sqlServers.Entities` var?
- Throw a different type of exception?

BTW I'm interested to learn about how you've set up your DBs in your project - assuming there's a primary and replica?

**Comment by abelal83** [comment]
Yep we log that also

```log
ValidateServerIsAllowedForRequest: SqlServerModel composed - SqlServerName xxx-xxx-sql-ne,ResourceGroup xxx-xxx-we, SubscriptionId xxxxx-xxxxb, TenantId xxxxx, ApprovedFirewallIps System.String[]
```

We've thrown different exceptions such as UnauthorizedException and same behavior, works locally but not in docker.

The DBs are setup in Azure using failover group, primary hosted in West and Secondary in North Europe. It's not Serenity related as such, just a managed SQL instance in Azure. We went with dacpac deployment instead of the db migrations provided by serenity, mainly because it was easier for us and as we're not planning on running outside of MS SQL it made logical sense.

**Reply by edwardch** [reply]
Yeah doesn't make sense to me either. 

So another check - could you try putting it into the `ValidateRequest()` override like so:

```cs
protected override void ValidateRequest()
{
    base.ValidateRequest();

    if (!isSqlServerPrimary)
    {
        // I'm using ValidationError here instead
        throw new ValidationError($" Request for the SQL Server {sqlServers.Entities[0].Name} is not possible via Serene beause the server is a secondary replica in readonly mode. Please contact Platform Engineering or DBAs for access.");
    }
}
```
___
And if still no luck, is there any chance you can set up a blank Serene project with the .NET 6 template, and add the same logic to see if it works?

Now the override is inside a file called YourClassSaveHandler.cs...

My Dockerfile (not optimised, but it works) looks like this:

```bash
FROM mcr.microsoft.com/dotnet/aspnet:6.0-bullseye-slim AS base
RUN apt-get update \
    && apt-get install -y --allow-unauthenticated \
        libc6-dev \
        libgdiplus \
        libx11-dev \
    && rm -rf /var/lib/apt/lists/*
WORKDIR /app
EXPOSE 80

FROM mcr.microsoft.com/dotnet/sdk:6.0-bullseye-slim AS build
RUN apt-get update -yq && \
    apt-get upgrade -yq \
    && apt-get install -yq curl git nano
RUN curl -sL https://deb.nodesource.com/setup_16.x | bash - && apt-get install -yq nodejs 
RUN npm install -g npm

WORKDIR /src
COPY ["NuGet.config", "."]
COPY ["Serene/Serene.Web/Serene.Web.csproj", "Serene/Serene.Web/"]
RUN dotnet restore --configfile NuGet.config  "Serene/Serene.Web/Serene.Web.csproj"
COPY . .
WORKDIR "/src/Serene/Serene.Web"
RUN dotnet build "Serene.Web.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "Serene.Web.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "Serene.Web.dll"]
```
___

> The DBs are setup in Azure using failover group, primary hosted in West and Secondary in North Europe. It's not Serenity related as such, just a managed SQL instance in Azure. 

Nice, same here!

> We went with dacpac deployment instead of the db migrations provided by serenity, mainly because it was easier for us and as we're not planning on running outside of MS SQL it made logical sense.

Sounds exactly like what I'd like to have set up. I'm using the db migrations which are now getting rather large (13Mb of procs/views/etc). Do you have any more information on your setup, or resources that you're able to share?

TIA

**Reply by abelal83** [reply]
The SQL setup is your usual Azure PaaS database with failover.
Dacpac was created with visual studio (not vs code), just open visual studio, create new project and add SQL Server Database project type
![image](https://user-images.githubusercontent.com/20558784/171599633-e5ba9bc5-c973-469d-9e3c-f0a6aa7f3318.png)

**Reply by edwardch** [reply]
> Dacpac was created with visual studio (not vs code), just open visual studio, create new project and add SQL Server Database project type

Cool, thank you! 

Our DB changes quite often, so thought you might be managing dacpac deployments in your CI/CD pipeline too.

**Comment by VictorTomaili** [comment]
Hi @abelal83 and @edwardch 
It's because the environment value is not the same in docker and in your local machine this value is `development`. When you release the application, your application must hide the details of the exception from the end user for security reason. But in your development machine you need to see all the details.

Check this
https://github.com/serenity-is/Serenity/blob/master/src/Serenity.Net.Web/Mvc/EndpointExtensions.cs#L17

And check this sample to customize your exceptions as you wish
https://github.com/serenity-is/Serene/blob/master/Serene/Serene.Web/Modules/Common/Helpers/HumanizeSqlExceptionBehavior.cs

**Reply by abelal83** [reply]
Ahh yes I recall that. I will test it out thank you @VictorTomaili

**Reply by abelal83** [reply]
That did it

for reference here is the docker command

```
docker run -v C:\src\appsettings.json:/app/appsettings.json --rm -it -p 5000:5000/tcp -e ASPNETCORE_ENVIRONMENT=Development serene-web:latest
```

Thank you both for your help!

**Reply by edwardch** [reply]
Ah, right, thanks Victor!

@abelal83 the ValidationError (or returning a different type of response) might also work instead of throwing an exception, or running production in Development mode.

**Reply by VictorTomaili** [reply]
Handling an exception like this by a behavior is a nice pattern for serenity. Don't expose the detailed exception message to the end user. And behaviors can run for all rows with the same connection key and/or handlers which have the interface with the `isSqlServerPrimary` property. With that, you can throw a custom exception for new and existing handlers without modify them.
